File: core/src/test/java/com/netflix/iceberg/TestSnapshotJson.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.netflix.iceberg;

import com.google.common.collect.ImmutableList;
import com.netflix.iceberg.exceptions.RuntimeIOException;
import com.netflix.iceberg.io.InputFile;
import com.netflix.iceberg.io.OutputFile;
import org.junit.Assert;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TemporaryFolder;
import java.io.File;
import java.io.IOException;
import java.util.List;

public class TestSnapshotJson {
  @Rule
  public TemporaryFolder temp = new TemporaryFolder();

  @Test
  public void testJsonConversion() {
    Snapshot expected = new BaseSnapshot(null, System.currentTimeMillis(),
        "file:/tmp/manifest1.avro", "file:/tmp/manifest2.avro");
    String json = SnapshotParser.toJson(expected);
    Snapshot snapshot = SnapshotParser.fromJson(null, json);

    Assert.assertEquals("Snapshot ID should match",
        expected.snapshotId(), snapshot.snapshotId());
    Assert.assertEquals("Files should match",
        expected.manifests(), snapshot.manifests());
  }

  @Test
  public void testJsonConversionWithManifestList() throws IOException {
    long parentId = 1;
    long id = 2;
    List<ManifestFile> manifests = ImmutableList.of(
        new GenericManifestFile("file:/tmp/manifest1.avro", 0),
        new GenericManifestFile("file:/tmp/manifest2.avro", 0));

    File manifestList = temp.newFile("manifests");
    Assert.assertTrue(manifestList.delete());
    manifestList.deleteOnExit();

    try (ManifestListWriter writer = new ManifestListWriter(
        Files.localOutput(manifestList), id, parentId)) {
      writer.addAll(manifests);
    }

    Snapshot expected = new BaseSnapshot(
        null, id, parentId, System.currentTimeMillis(), Files.localInput(manifestList));
    Snapshot inMemory = new BaseSnapshot(
        null, id, parentId, expected.timestampMillis(), manifests);

    Assert.assertEquals("Files should match in memory list",
        inMemory.manifests(), expected.manifests());

    String json = SnapshotParser.toJson(expected);
    Snapshot snapshot = SnapshotParser.fromJson(new LocalTableOperations(), json);

    Assert.assertEquals("Snapshot ID should match",
        expected.snapshotId(), snapshot.snapshotId());
    Assert.assertEquals("Timestamp should match",
        expected.timestampMillis(), snapshot.timestampMillis());
    Assert.assertEquals("Parent ID should match",
        expected.parentId(), snapshot.parentId());
    Assert.assertEquals("Manifest list should match",
        expected.manifestListLocation(), snapshot.manifestListLocation());
    Assert.assertEquals("Files should match",
        expected.manifests(), snapshot.manifests());
  }

  private class LocalTableOperations implements TableOperations {
    @Override
    public TableMetadata current() {
      throw new UnsupportedOperationException("Not implemented for tests");
    }

    @Override
    public TableMetadata refresh() {
      throw new UnsupportedOperationException("Not implemented for tests");
    }

    @Override
    public void commit(TableMetadata base, TableMetadata metadata) {
      throw new UnsupportedOperationException("Not implemented for tests");
    }

    @Override
    public InputFile newInputFile(String path) {
      return Files.localInput(path);
    }

    @Override
    public OutputFile newMetadataFile(String filename) {
      try {
        File metadataFile = temp.newFile(filename);
        metadataFile.delete();
        metadataFile.deleteOnExit();
        return Files.localOutput(metadataFile);
      } catch (IOException e) {
        throw new RuntimeIOException(e);
      }
    }

    @Override
    public void deleteFile(String path) {
      new File(path).delete();
    }

    @Override
    public long newSnapshotId() {
      throw new UnsupportedOperationException("Not implemented for tests");
    }
  }
}

File: python/tests/core/test_snapshot_json.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

from iceberg.api import DataOperations
from iceberg.core import BaseSnapshot, SnapshotParser


def test_json_conversion(ops, expected_base_snapshot):
    as_json = SnapshotParser.to_json(expected_base_snapshot)
    snapshot = SnapshotParser.from_json(ops, as_json)

    assert expected_base_snapshot.snapshot_id == snapshot.snapshot_id
    assert expected_base_snapshot.manifests == snapshot.manifests


def test_json_conversion_with_operation(ops, snapshot_manifests):
    parent_id = 1
    id = 2
    expected = BaseSnapshot(ops=ops, snapshot_id=id, parent_id=parent_id,
                            manifests=snapshot_manifests, operation=DataOperations.REPLACE,
                            summary={"files-added": 4,
                                     "files-deleted": "100"})

    json_obj = SnapshotParser.to_json(expected)
    snapshot = SnapshotParser.from_json(ops, json_obj)

    assert expected.snapshot_id == snapshot.snapshot_id
    assert expected.timestamp_millis == snapshot.timestamp_millis
    assert expected.parent_id == snapshot.parent_id
    assert expected.manifest_location == snapshot.manifest_location
    assert expected.manifests == snapshot.manifests
    assert expected.operation == snapshot.operation
    assert expected.summary == snapshot.summary

# def test_conversion_with_manifest_list(snapshot_manifests):
#     parent_id = 1
#     id = 2
#
#     with NamedTemporaryFile() as manifest_list:
#         with ManifestListWriter(Files.local_output(manifest_list), id, parent_id) as writer:
#             writer.add_all(snapshot_manifests)
#
#         with open(manifest_list) as fo:
#             print(fo.read())
